client/src/main/java/io/pravega/client/stream/impl/Pinger.java

```diff
@@ -87,8 +88,8 @@ private long getPingInterval(long txnLeaseMillis) {
     private void startPeriodicPingTxn() {
         if (!isStarted.get()) {
             log.info("Starting Pinger at an interval of {}ms ", this.pingIntervalMillis);
-            executor.scheduleAtFixedRate(this::pingTransactions, 10, this.pingIntervalMillis,
-                    TimeUnit.MILLISECONDS);
+            // scheduleAtFixedRate ensure that there are no concurrent executions of the command, pingTransactions()
+            scheduledFuture = executor.scheduleAtFixedRate(this::pingTransactions, 10, this.pingIntervalMillis, TimeUnit.MILLISECONDS);
```

Comment:
`scheduledFuture` is not guarded by anything so it's not thread safe. 

If you make this `AtomicReference<ScheduledFuture>` then you will have solved that and eliminate the need for `isStarted` (you can just check if this object has anything in it).

Reply:
Used `AtomicReference` for the `ScheduledFuture` and used `AtomicBoolean isStarted` to remove @Synchronized.